src/katgpucbf/xbgpu/kernels/beamform.mako

```diff
@@ -0,0 +1,120 @@
+/*******************************************************************************
+ * Copyright (c) 2023, National Research Foundation (SARAO)
+ *
+ * Licensed under the BSD 3-Clause License (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy
+ * of the License at
+ *
+ *   https://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ ******************************************************************************/
+
+<%include file="/port.mako"/>
+
+#define N_BEAMS ${len(beam_pols)}
+#define BLOCK_TIME ${block_time}
+#define BLOCK_CHANNELS ${block_channels}
+#define QMAX 127
+
+// Quantise, and return saturation flag
+// TODO: this is copied from postproc.py. Create a shared implementation.
+DEVICE_FN bool quant(float value, int *out)
+{
+    float clamped = fminf(fmaxf(value, -QMAX), QMAX);
+    bool saturated = value != clamped;
+
+    int q;
+    // Convert to s8, round to nearest integer, and saturate
+    // (saturation is redundant due to the test above, but automatic
+    // for float-to-int conversions on CUDA hardware).
+#ifdef __OPENCL_VERSION__
+    q = convert_char_sat_rte(clamped);
+#else
+    asm("cvt.rni.sat.s8.f32 %0, %1;" : "=r" (q) : "f"(clamped));
+#endif
+    *out = q;
+    return saturated;
+}
+
+// Compute a * b in complex numbers
+DEVICE_FN float2 cmul(float2 a, float2 b)
+{
+    return make_float2(a.x * b.x - a.y * b.y, a.x * b.y + a.y * b.x);
+}
+
+// Compute a * b + c in complex numbers
+DEVICE_FN float2 cmad(float2 a, float2 b, float2 c)
+{
+    return make_float2(a.x * b.x - a.y * b.y + c.x, a.x * b.y + a.y * b.x + c.y);
+}
+
+// out: shape frame, beam, channel, time
+// in: shape frame, antenna, channel, time, pol
+// weight: shape antenna, beam, tightly packed
+// delays: shape antenna, beam, tightly packed
+// Each thread computes all beams for one (channel, time)
+KERNEL REQD_WORK_GROUP_SIZE(BLOCK_TIME, BLOCK_CHANNELS, 1) void beamform(
+    GLOBAL char2 * RESTRICT out,
+    GLOBAL const char4 * RESTRICT in,
+    GLOBAL const float2 * RESTRICT weights,
+    GLOBAL const float * RESTRICT delays,
+    int out_stride,
+    int out_beam_stride,
+    int out_frame_stride,
+    int in_stride,
+    int in_antenna_stride,
+    int in_frame_stride,
+    int n_antennas,
+    int n_channels,
+    int n_times
+)
+{
+    int time = get_global_id(0);
+    int channel = get_global_id(1);
+    int frame = get_group_id(2);
```

Comment:
I think one can go too far with trying to be globally consistent. In this context there isn't much ambiguity, since there aren't other things that these could be confused with.